### ISSUE TITLE
Add scenario to deploy a 'realistic' workload: Cool Store

### ISSUE THREAD

**Issue by gbraad** [minishift/minishift#964]
Implement a scenario to deploy a workload that contains a frontend, API gateway and some services, which represent a realistic workload.

**Comment by gbraad** [comment]
```cucumber
@coolstore
Feature: Cool Store
  In order to test Minishift under load
  I need to setup a test environment
  Using CentOS as image and 4G of memory
  After which I will import image streams
  Finally I will deploy the Cool Store
  During this deployment I will measure memory and CPU usage


  Scenario: Starting Minishift
    Given Minishift has state "Does Not Exist"
     When executing "minishift start --iso-url https://github.com/minishift/minishift-centos-iso/releases/download/v1.0.0/minishift-centos7.iso --memory 4096" succeeds
     Then Minishift should have state "Running"

  Scenario: Import JBoss imagestreams
    Given Minishift has state "Running"
     When executing "oc --as system:admin create -n openshift -f https://raw.githubusercontent.com/jboss-openshift/application-templates/master/jboss-image-streams.json" succeeds
     Then stdout should contain
      """

      """
      
  Scenario: Import JBoss imagestreams
    Given Minishift has state "Running"
     When executing "oc --as system:admin create -n openshift -f https://raw.githubusercontent.com/jboss-fuse/application-templates/master/fis-image-streams.json" succeeds
     Then stdout should contain
      """

      """

  Scenario: User can create new namespace coolstore
   Given Minishift has state "Running"
    When executing "oc new-project coolstore" succeeds
    Then stdout should contain
     """
     Now using project "coolstore"
     """
  
  Scenario: User can deploy application to namespace coolstore 
   Given Minishift has state "Running"
    When executing "oc process -f https://raw.githubusercontent.com/jbossdemocentral/coolstore-microservice/1.0.x/openshift/coolstore-template.yaml | oc create -f -" succeeds
    Then stdout should contain
     """
     Success
     """

  Scenario: Successfully built the cart service 
   Given Minishift has state "Running" 
    When executing "oc bc/cart --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully deployed cart service to namespace coolstore 
   Given Minishift has state "Running" 
    When executing "oc rollout status deploymentconfig cart --watch" succeeds
    Then stdout should contain
     """
     "cart" successfully rolled out
     """

  Scenario: Successfully rolled out gateway to namespace coolstore 
   Given Minishift has state "Running" 
    When executing "oc rollout status deploymentconfig coolstore-gw --watch" succeeds
    Then stdout should contain
     """
     "coolstore-gw" successfully rolled out
     """

  Scenario: Successfully built the gateway 
   Given Minishift has state "Running" 
    When executing "oc bc/coolstore-gw --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully rolled out gateway to namespace coolstore 
   Given Minishift has state "Running" 
    When executing "oc rollout status deploymentconfig coolstore-gw --watch" succeeds
    Then stdout should contain
     """
     "coolstore-gw" successfully rolled out
     """

  Scenario: Able to interact with the catalog from the coolstore gateway 
   Given Minishift has state "Running" 
    When executing "oc rsh $(oc get pods -o name -l application=coolstore-gw) curl http://catalog:8080/api/products" succeeds
    Then stdout should contain
     """
     Fedora
     """

  Scenario: Able to interact with the inventory from the coolstore gateway 
   Given Minishift has state "Running" 
    When executing "oc rsh $(oc get pods -o name -l application=coolstore-gw) curl http://inventory:8080/api/availability/329299" succeeds
    Then stdout should contain
     """
     Raleigh
     """

  Scenario: Able to interact with the cart from the coolstore gateway 
   Given Minishift has state "Running" 
    When executing "oc rsh $(oc get pods -o name -l application=coolstore-gw) curl http://cart:8080/api/cart/FOO" succeeds
    Then stdout should contain
     """
     cartTotal
     """

  Scenario: Stopping Minishift
    Given Minishift has state "Running"
     When executing "minishift stop" succeeds
     Then Minishift should have state "Stopped"

  Scenario: Deleting Minishift
    Given Minishift has state "Stopped"
     When executing "minishift delete" succeeds
     Then Minishift should have state "Does Not Exist"
     When executing "minishift ip"
     Then exitcode should equal 1
```

**Comment by LalatenduMohanty** [comment]
@gbraad we should  also run helloworld-msa  in integration tests as we have people demoing it on Minishift/CDK. So we should make sure it always on Minishift.  

The only drawback is that the execution time of integration tests will take more time (significantly) with this.

**Comment by hferentschik** [comment]
@gbraad Nice, does this work already? That would be awesome.

**Comment by gbraad** [comment]
Wasn't able to fully test it. At the moment it runs till:

```cucumber
    When  executing "oc process -f https://raw.githubusercontent.com/jbossdemocentral/coolstore-microservice/1.0.x/openshift/coolstore-template.yaml | oc create -f -" succeeds # features/coolstore.feature:42
      Error: Command did not execute successfully. cmdExit: 1, cmdErr: error: template name must be specified only once: oc
See 'oc process -h' for help and examples.
```

**Comment by gbraad** [comment]
It fails on: https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/process.go#L126 (only when using `godog`)

**Comment by gbraad** [comment]
Finally understand WHY this command fails. Of course `RunCommand` is not a functional shell, so I already replaced the `|`, but of course `>` also does not work. These are all shell related functionality. At the moment this gets passed to the `oc process` command and is interpreted as a command line parameter. Another issue we will certainly run into is related to the fact of using `$( )` or backticks or setting an export variable, as can be seen in the `oc rsh` commands.

**Comment by gbraad** [comment]
This task is now mostly about adding functionality to the `godog`-based integration tests to be able to deal with more complex scenarios, such as the one above. Some of the functionality is focussed on allowed variables to be used integration:

  * Especially as `oc process` can not write to a file (depends on piping)
  * and that `oc rsh` needs a podname (these are dynamic)

After this, I will also need to dump results in a file (which we can also use for debug/error purposes I hope)

The cucumber could look like:

```cucumber
When executing "oc get pods" store as "ALLPODS"
Then variable "ALLPODS" should contain "cart-1"
```

```cucumber
When executing "oc get pods -o name -l application=api-gw" store as "APIGW"
Then executing "oc rsh [APIGW] curl http://cart/FOO" should contain "cartTotal: 1"
```

This way we can re-use return values and perform further calls using these values
The first case of contains allows to perform multiple checks on the ALLPODS which you can not do now
and the second case allows you do perform re-use of the value in multiple calls for instance. I need to make several calls on the APIGW to verify it can reach each deployed service

Already discussion with @agajdosi if functionality is missing for a similar to add more test scenarios: https://issues.jboss.org/browse/CDK-131

**Comment by LalatenduMohanty** [comment]
> This way we can re-use return values and perform further calls using these values
> The first case of contains allows to perform multiple checks on the ALLPODS which you can not do now
> and the second case allows you do perform re-use of the value in multiple calls for instance. I need to make several calls on the APIGW to verify it can reach each deployed service
> 
> Already discussion with @agajdosi if functionality is missing for a similar to add more test scenarios: https://issues.jboss.org/browse/CDK-131

+1

**Comment by agajdosi** [comment]
> The only drawback is that the execution time of integration tests will take more time (significantly) with this.

I would suggest to add `@openshift` tag to coolstore feature and then to MSA hello world feature in future and not tag them with `@basic` to exclude them from `make integration` which currently runs only those tests, which have `@basic` tag. Following this plan `make integration` would stay relatively quick check for developers to assure that basic functionality is working, while they could specify `@openshift` to run more specific set of tests when they work on OpenShift related issues - for now coolstore, later MSA hello world or more.

What do you think of this solution, @gbraad @LalatenduMohanty?

This is related to #583. So I can be the one who will add those tags after PR is merged, also if we agree on `@openshift` tag I will add these tags to basic.feature to all scenarios which deploy sample application on OpenShift, so `make integration GODOG_OPTS= '-tags openshift'` will run coolstore.feature and also deployment of this sample application.

**Comment by LalatenduMohanty** [comment]
@agajdosi +1. IMO we should run the @opeshidt tests before each release.

**Comment by agajdosi** [comment]
@LalatenduMohanty agree, for downstream I am planning for every night on builds of CDK

**Comment by gbraad** [comment]
@agajdosi it was already tagged as `@coolstore` for this purposes. Consider them extended tests... just wondering why you chose the tag `@openshift`, as I believe everything we do is kind of related to OpenShift ;-). I'd rather allow for more fine-grained specification... WDYT?

**Comment by gbraad** [comment]
Removed the `oc process -f | oc apply -f -` (this way no `stdin` workaround is needed

```cucumber
@coolstore
Feature: Cool Store
  In order to test Minishift under load
  I need to setup a test environment
  Using CentOS as image and 4G of memory
  After which I will import image streams
  Finally I will deploy the Cool Store
  During this deployment I will measure memory and CPU usage


  Scenario: Starting Minishift
    Given Minishift has state "Does Not Exist"
     When executing "minishift start --iso-url https://github.com/minishift/minishift-centos-iso/releases/download/v1.0.0/minishift-centos7.iso --memory 4096" succeeds
     Then Minishift should have state "Running"

  Scenario: Import JBoss imagestreams
    Given Minishift has state "Running"
     When executing "oc --as system:admin create -n openshift -f https://raw.githubusercontent.com/jboss-openshift/application-templates/master/jboss-image-streams.json" succeeds
     Then stdout should contain
      """
      imagestream "redhat-openjdk18-openshift" created
      """
      
  Scenario: Import JBoss imagestreams
    Given Minishift has state "Running"
     When executing "oc --as system:admin create -n openshift -f https://raw.githubusercontent.com/jboss-fuse/application-templates/master/fis-image-streams.json" succeeds
     Then stdout should contain
      """
      imagestream "fis-java-openshift" created
      """

  Scenario: User can create new namespace coolstore
   Given Minishift has state "Running"
    When executing "oc new-project coolstore" succeeds
    Then stdout should contain
     """
     Now using project "coolstore"
     """
  
  Scenario: User can create a template for coolstore 
   Given Minishift has state "Running"
    When executing "oc create -f https://raw.githubusercontent.com/jbossdemocentral/coolstore-microservice/1.0.x/openshift/coolstore-template.yaml" succeeds
    Then stdout should contain
     """
     template "coolstore" created
     """

  Scenario: User can deploy the coolstore template to namespace coolstore 
   Given Minishift has state "Running"
    When executing "oc new-app --template=coolstore" succeeds
    Then stdout should contain
     """
     Success
     """

  Scenario: Successfully built the Web UI service 
   Given Minishift has state "Running" 
    When executing "oc logs bc/web-ui --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully built the inventory service 
   Given Minishift has state "Running" 
    When executing "oc logs bc/inventory --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully built the catalog service 
   Given Minishift has state "Running" 
    When executing "oc logs bc/catalog --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully built the cart service 
   Given Minishift has state "Running" 
    When executing "oc logs bc/cart --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully deployed cart service to namespace coolstore 
   Given Minishift has state "Running" 
    When executing "oc rollout status deploymentconfig cart --watch" succeeds
    Then stdout should contain
     """
     "cart" successfully rolled out
     """

  Scenario: Successfully built the gateway 
   Given Minishift has state "Running" 
    When executing "oc logs bc/coolstore-gw --follow" succeeds
    Then stdout should contain
     """
     Push successful
     """

  Scenario: Successfully rolled out gateway to namespace coolstore 
   Given Minishift has state "Running" 
    When executing "oc rollout status deploymentconfig coolstore-gw --watch" succeeds
    Then stdout should contain
     """
     "coolstore-gw" successfully rolled out
     """

  Scenario: Successfully deployed all services
   Given Minishift has state "Running" 
    When executing "oc status -v" succeeds
    Then stdout should not contain
     """
     has failed
     """

  Scenario: Able to interact with the catalog from the coolstore gateway 
   Given Minishift has state "Running" 
    When set variable "apigw" executing "oc get pods -o name -l application=coolstore-gw"
     And executing "oc rsh $(apigw) curl http://catalog:8080/api/products" succeeds
    Then stdout should contain
     """
     Fedora
     """

  Scenario: Able to interact with the inventory from the coolstore gateway 
   Given Minishift has state "Running"
    When set variable "apigw" executing "oc get pods -o name -l application=coolstore-gw"
     And executing "oc rsh $(apigw) curl http://inventory:8080/api/availability/329299" succeeds
    Then stdout should contain
     """
     Raleigh
     """

  Scenario: Able to interact with the cart from the coolstore gateway 
   Given Minishift has state "Running"
    When set variable "apigw" executing "oc get pods -o name -l application=coolstore-gw"
     And executing "oc rsh $(apigw) curl http://cart:8080/api/cart/FOO" succeeds
    Then stdout should contain
     """
     cartTotal
     """

  Scenario: Stopping Minishift
    Given Minishift has state "Running"
     When executing "minishift stop" succeeds
     Then Minishift should have state "Stopped"

  Scenario: Deleting Minishift
    Given Minishift has state "Stopped"
     When executing "minishift delete" succeeds
     Then Minishift should have state "Does Not Exist"
     When executing "minishift ip"
     Then exitcode should equal 1
```

**Comment by praveenkumar** [comment]
@gbraad Since now xpaas addon contain required imagestreams, shouldn't we run minishift with this addon enabled so we can remove imagestream details? At least once we have this addon as part of default so it can be installed and enabled.

**Comment by gbraad** [comment]
@praveenkumar if enabled by default, sure...

**Comment by gbraad** [comment]
Not sure what happened here ...
```
......................F-F--F--F--F--F--F--F---F---F---F--F---- 62


--- Failed steps:

    When  executing "oc logs bc/inventory --follow" succeeds # features/coolstore.feature:66
      Error: Command 'logs bc/inventory --follow' did not execute successfully. cmdExit: 1, cmdErr: error: read tcp 192.168.42.1:37892->192.168.42.13:8443: read: connection timed out


    Given  Minishift has state "Running" # features/coolstore.feature:73
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:81
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:89
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:97
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:105
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:113
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:121
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:130
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:139
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Running" # features/coolstore.feature:148
      Error: Minishift state did not match. Expected: Running, Actual: Paused

    Given  Minishift has state "Stopped" # features/coolstore.feature:153
      Error: Minishift state did not match. Expected: Stopped, Actual: Paused
```

**Comment by hferentschik** [comment]
> Since now xpaas addon contain required imagestreams, shouldn't we run minishift with this addon enabled

+1

**Comment by hferentschik** [comment]
>  When executing "oc logs bc/coolstore-gw --follow" succeeds

I am not sure whether this is going to work all the time. The problem is that there is a timing issue and the log command might not successfully tail after a new-app. This was one of the reasons why it used to be so hard to write a reliable script to wait for a build to complete. IMO we should remove the _oc logs_ commands and just use _oc rollout_. AFAIK this is the right way of doing it.

**Comment by agajdosi** [comment]
> just wondering why you chose the tag @openshift, as I believe everything we do is kind of related to OpenShift ;-)

I agree. OpenShift is the reason why all in Minishift is happening, but there are also plenty of stuff on this way to have OpenShift running locally and that was my perspective - we can have `@vm`, `@ux`, `@error-reporting`... and also `@openshift` (not sure with the name on 100% - maybe it could be `@app-deployment` or something more accurate). More and more I am starting to think of `@oc` as this set of scenarios will test functionality of OpenShift and this it will use "oc" command.

> I'd rather allow for more fine-grained specification...

It is possible to be fine-grained with: `GODOG_OPTS="-paths /home/<username>/go/src/github.com/minishift/minishift/test/integration/features/coolstore.feature"`. But as there is no limit on tags, I agree with `@coolstore`. And as part of #583 I think I can later add second tag which will group all scenarios which focus on functionality of OpenShift, I will be thankful for any naming preference/suggestion :) WDYT, @gbraad?

**Comment by gbraad** [comment]
@agajdosi locally I can run the test on b2d. But as soon as I use the CentOS image, I run into issues...

At the moment it seems the `set variable`-step is faulty and does not assign any value.

**Comment by hferentschik** [comment]
merged via pull request #986
